api/python/t4/bucket.py

```diff
@@ -0,0 +1,221 @@
+            meta(dict): optional user-provided metadata to store
+        """
+        dest = self._uri + key
+        meta = meta or {}
+        data, target = serialize_obj(obj)
+        all_meta = dict(
+            target=target.value,
+            user_meta=meta
+        )
+        put_bytes(data, fix_url(dest), all_meta)
+
+    def put_file(self, key, path):
+        """
+        Stores file at path to key in bucket.
+
+        Args:
+            key(str): key in bucket to store file at
+            path(str): string representing local path to file
+
+        Returns:
+            None
+
+        Raises:
+            if no file exists at path
+            if copy fails
+        """
+        dest = self._uri + key
+        copy_file(path, dest)
+
+    def put_dir(self, key, directory):
+        """
+        Stores all files under directory under the prefix key.
+
+        Args:
+            key(str): prefix to store files under in bucket
+            directory(str): path to local directory to grab files from
+
+        Returns:
+            None
+
+        Raises:
+            if directory isn't a valid local directory
+            if writing to bucket fails
+        """
+        # Ensure key ends in '/'.
+        if key[-1] != '/':
+            key = key + '/'
+
+        src_path = pathlib.Path(directory)
+        if not src_path.is_dir():
+            raise QuiltException("Provided directory does not exist")
+
+        files = src_path.rglob('*')
+        for f in files:
+            if not f.is_file():
+                continue
+            new_key = key + f.relative_to(src_path).as_posix()
+            new_path = self._uri + new_key
+            copy_file(f.resolve().as_uri(), new_path)
+
+    def keys(self):
+        """
+        Lists all keys in the bucket.
+
+        Returns:
+            list of strings
+        """
+        return list(map(lambda x: x.get('Key'), list_objects(self._bucket, '')))
+
+    def delete(self, key):
+        """
+        Deletes a key from the bucket.
+
+        Args:
+            key(str): key to delete
+
+        Returns:
+            None
+
+        Raises:
+            if delete fails
+        """
+        delete_object(self._bucket, key)
+
+    def fetch(self, key, path):
+        """
+        Fetches file (or files) at key to path.
+
+        If key ends in '/', then all files with the prefix key will match and will
+            be stored in a directory at path.
+        Otherwise, only one file will be fetched and it will be stored at path.
+
+        Args:
+            key(str): key in bucket to fetch
+            path(str): path in local filesystem to store file or files fetched
+
+        Returns:
+            None
+
+        Raises:
+            if path doesn't exist
+            if download fails
+        """
+        if not key.endswith('/'):
+            copy_file(self._uri + key, fix_url(path))
+            return
+
+        _download_dir(self._bucket, key, fix_url(path))
```

Comment:
Just use `copy_file` for everything - it already checks for `/` and downloads dirs.

You can also use `download_file` to avoid dealing with URLs.